06_pointers_and_slices.md

```diff
@@ -1,27 +1,65 @@
-
-In Rust, if a data type is *dynamically sized*, meaning it can expand or shrink and its size is not known at compile time, it *must be* allocated on the heap. If the data is a known, fixed-size and doesn't change, it can be allocated on the stack. This is for more memory safety purposes. Remember, the stack memory is limited so the compiler doesn't want there to be any surprises when the program actually runs which could lead to dangerous memory errors. Since the pointers themselves are known and fixed in size they can be safely allocated to our program's stack. The underlying data they point to is allocated on the heap.
-
-Notice in the diagram above how the vector is also a pointer type to data stored on the heap. In Rust, the vector is actually just a *smart pointer*, unlike the slice, which is instead of a *fat pointer*. A smart pointer contains additional metadata and capabilities. It also *owns* the data instead of just borrowing a reference to it. We'll explore the concepts of borrowing and references in more detail later on. For now, it's enough to understand the following key points:
-1. Both vectors and slice references (often just called "slices" for short) **point** to the data in memory. This makes it lightweight to pass around and move these data types in the program. When they are moved, there is no need to move or copy the data on the heap as well. 
-2. A vector indicates ownership of the memory and a slice indicates a borrowing of memory. One way to think about this is that when the vector goes out of scope and is no longer used or is "dropped", it has to deallocate all the data in memory as well. So when the smart pointer is removed, all the underlying data on the heap must be removed as well. The slice reference however can be "dropped" and the data it points to will remain in heap memory, since it is just borrowing the memory and doesn't own it.
-
-Hopefully that made sense. Don't worry if it feels a bit too complicated at the moment. We'll get more familiar with these concepts as we progress in the course and get more practice. It will eventually make sense. I promise. 
+If you don't work in systems programming, you probably don't spend much time thinking about the stack and the heap, so let's provide a quick refresher here.
+The stack is a region in memory that holds local variables during our program execution.
+Those variables in turn can refer to, or *point to*, data on the heap which is a less structured area of memory available to our program.
+When we need to store large amounts of data, we typically *allocate* that data on the heap.
+This is useful because the heap has no memory restrictions, whereas the stack is limited.
+The heap also allows data to be accessed from anywhere in the program, which is useful for data shared across different functions or modules.
```

Comment:
Because the stack is generally more constrained than the heap and has a fixed size from my understanding

Reply:
Would you prefer the following:

"This is useful because the heap has no memory restrictions, where as the stack is generally more constrained"

Reply:
> Because the stack is generally more constrained than the heap and has a fixed size from my understanding

I understand that, in practice, modern OSes can limit the stack size for programs. But this is not the mental model we have in mind when programming.

Reply:
> Would you prefer the following:
> 
> "This is useful because the heap has no memory restrictions, where as the stack is generally more constrained"

What about suppressing line 18 altogether? Rereading the text, it adds nothing indeed.

Reply:
Why does Rust enforce the rule that the size of types on the stack must be known at compile time? Are you suggesting it is for memory efficiency purposes as opposed to memory safety?

Reply:
I think its important to articulate the tradeoff between the stack and the heap. From my understanding, the stack is generally more constrained and less flexible because of how its structured, but more efficient. But this understanding might be wrong as you're suggesting. 

How would you articulate the tradeoff?

Reply:
> Why does Rust enforce the rule that the size of types on the stack must be known at compile time? Are you suggesting it is for memory efficiency purposes as opposed to memory safety?

It's reminiscent of C and before. The stack is used to store local variables. So, imagine we save a local variable (say x_var) in the stack and enter a new local scope e.g. by calling a function. In this scope we may have access to x_var via a pointer. If we use this pointer to try to expand (or shrink) the memory required to hold x_var data, we would need to unwrap and rebuild the whole stack. This is technically possible, but would degrade performance a lot, and we risk leaking the stack abstraction to the user of the language. So, we impose the discipline of fixed-size-only for data in the stack.

I agree this is a limitation of the stack. It's just that we usually refer to it that way: the stack is used for a specific task and the discipline imposed gives us some performance guarantees. Everything else goes to the heap that has essentially no discipline, but also no guarantees for optimization.

Reply:
> Ok how about the following for the two paragraphs (removing the second mention of "limited" as well):
> 
> "When we need to store large amounts of data or a data type whose size can change during the runtime of our program, we typically _allocate_ that data on the heap. The heap is more flexible and less memory-constrained than the stack. The heap also allows data to be accessed from anywhere in the program, which is useful for data shared across different functions or modules. However, allocating to the heap comes with a cost...
> 
> Rust enforces that any type that is _dynamically sized_, meaning it can expand or shrink and its size is not known at compile time, must be allocated to the heap. If the data size is known and fixed at compile time, and doesn't change during execution, it can be allocated on the stack. This is for both memory safety and memory efficiency purposes.
> 
> Since pointers themselves are known and fixed in size they can be safely placed on our program's stack. The underlying data they point to are allocated on the heap."

Added this to the commit, covers the idea of stack and heap and sound straightforward now.

Reply:
Ah thanks for the explanation! That makes more sense to me now. Good thing there's a professor of computer science going through this course material! 🙌 😅